File: service/app/src/test/java/io/miragon/blueprint/process/util/TimerUtils.java
package io.miragon.blueprint.process.util;

import io.miragon.bpmn.runtime.ElementId;
import org.operaton.bpm.engine.ProcessEngine;
import org.operaton.bpm.engine.runtime.Job;

public final class TimerUtils {

    private TimerUtils() {
    }

    /**
     * Fires the timer job of the given boundary/catch event directly, regardless of its due date.
     * Replaces clock manipulation: the tests verify that the timer path is wired correctly, not the
     * real-world waiting duration.
     */
    public static void fireTimer(ProcessEngine engine, ElementId timerActivityId) {
        Job timer = engine
            .getManagementService()
            .createJobQuery()
            .timers()
            .activityId(timerActivityId.getValue())
            .singleResult();
        if (timer == null) {
            throw new IllegalArgumentException("no timer job found for activity '" + timerActivityId.getValue() + "'");
        }
        engine.getManagementService().executeJob(timer.getId());
    }
}

File: service/app/src/test/kotlin/io/miragon/blueprint/process/util/TimerUtils.kt
package io.miragon.blueprint.process.util

import io.miragon.bpmn.runtime.ElementId
import org.operaton.bpm.engine.ProcessEngine

/**
 * Fires the timer job of the given boundary/catch event directly, regardless of its due date.
 * Replaces clock manipulation: the tests verify that the timer path is wired correctly, not the
 * real-world waiting duration.
 */
fun ProcessEngine.fireTimer(timerActivityId: ElementId) {
    val timer =
        managementService
            .createJobQuery()
            .timers()
            .activityId(timerActivityId.value)
            .singleResult()
    requireNotNull(timer) { "no timer job found for activity '${timerActivityId.value}'" }
    managementService.executeJob(timer.id)
}
